File: rotators/aws-lambda/src/postgres.ts
import pg from "pg"

export interface PostgresRotateParams {
    type: "postgres";
    host: string;
    port: number;
    database: string;

    managingUser: PostgresUser;
    rotateUser: PostgresUser;
}

interface PostgresUser {
    username: string;
    password: string;
    newPassword?: string;
}

export default async function rotate(event: PostgresRotateParams) {
    let statusCode = 200;
    let message = "ok";

    const { database, host, port, managingUser, rotateUser } = event;
    const { username: managingUsername, password: managingPassword } = managingUser;
    const { username: rotateUsername, newPassword: newPassword } = rotateUser;

    const { Client, escapeIdentifier, escapeLiteral } = pg;

    if (!newPassword) {
        return {
            code: 400,
            message: "No new password provided",
        }
    }

    try {
        const client = new Client({
            user: managingUsername,
            password: managingPassword,
            host: host,
            port: port,
            database: database,
        })
        await client.connect();

        const query = `ALTER USER ${escapeIdentifier(rotateUsername)} WITH PASSWORD ${escapeLiteral(newPassword)}`;
        await client.query(query);

        await client.end();
    } catch (error) {
        statusCode = 400;
        message = `Error rotating user '${rotateUsername}': ${error}`;
    }

    return {
        code: statusCode,
        message: message,
    }
}

File: rotators/aws-lambda/postgres/postgres.go
package postgres

import (
	"context"
	"database/sql"
	"fmt"

	"github.com/lib/pq"
)

type PostgresRotateParams struct {
	Host     string `json:"host"`
	Port     int    `json:"port"`
	Database string `json:"database"`

	ManagingUser PostgresUser `json:"managingUser"`
	RotateUser   PostgresUser `json:"rotateUser"`
}

type PostgresUser struct {
	Username    string  `json:"username"`
	Password    string  `json:"password"`
	NewPassword *string `json:"newPassword"`
}

func Rotate(ctx context.Context, request PostgresRotateParams) error {
	db, err := sql.Open("postgres", fmt.Sprintf("postgres://%s:%s@%s:%d/%s?sslmode=disable",
		request.ManagingUser.Username,
		request.ManagingUser.Password,
		request.Host,
		request.Port,
		request.Database,
	))
	if err != nil {
		return err
	}
	defer db.Close()

	if request.RotateUser.NewPassword == nil {
		return fmt.Errorf("no password provided")
	}

	_, err = db.ExecContext(ctx, fmt.Sprintf(`ALTER USER %s WITH PASSWORD %s`,
		pq.QuoteIdentifier(request.RotateUser.Username),
		pq.QuoteLiteral(*request.RotateUser.NewPassword)),
	)
	if err != nil {
		return fmt.Errorf("error rotating user %q: %w", request.RotateUser.Username, err)
	}
	return nil
}
